playlist.go

```diff
@@ -1,37 +1,58 @@
 package main
 
 import (
-	"fmt"
+	"sort"
 
 	"github.com/zmb3/spotify"
 )
 
-type playlist struct {
-	tracks []spotify.FullTrack
+type PlaylistService struct {
+	RelatedArtistService *RelatedArtistService
 }
 
-func generateFromRelatedArtists(ra []relatedartist) playlist {
+func NewPlaylistService(relatedArtistService *RelatedArtistService) *PlaylistService {
+	return &PlaylistService{
+		RelatedArtistService: relatedArtistService,
+	}
+}
 
-	var tracks []spotify.FullTrack
+func (ps *PlaylistService) GeneratePlaylist(trackID string, targetPopularity, threshold int) ([]spotify.FullTrack, error) {
+	artist := ps.RelatedArtistService.getFirstRelatedArtistByTrackId(trackID)
+	allTopTracks, err := ps.RelatedArtistService.getTopTracksFromRelatedArtists(artist.id, 1)
+	if err != nil {
+		return nil, err
+	}
 
-	for _, relatedartist := range ra {
-		topTrack := relatedartist.topTracks[0]
+	sort.Sort(byPopularity(allTopTracks))
 
-		tracks = append(tracks, topTrack)
-	}
+	similarTracks := ps.chooseSimilarPopularity(allTopTracks, targetPopularity, threshold)
+	uniqueTracks := ps.removeDuplicates(similarTracks)
 
-	return playlist{
-		tracks: tracks,
-	}
+	return uniqueTracks, nil
 }
 
-func displayPlaylist(p playlist) string {
+func (ps *PlaylistService) chooseSimilarPopularity(tracks []spotify.FullTrack, targetPopularity, threshold int) []spotify.FullTrack {
```

Comment:
i guess it's an arbitrary property that you can easily use to refine a search, it could be swapped in for another value held against a song if a good one existed... i guess the downside to it is that you then end up with a playlist full of songs with similar popularity, but the idea is you can then change how varied you go in this area, meaning if you want to generate a playlist with some similarly mainstream songs, and some niche songs, then you can widen the net. i would say it's not ideal though, and more an experiment in what results it produces than anything else.